src/parser/cvc/Cvc.g

```diff
@@ -880,6 +889,69 @@ mainCommand[std::unique_ptr<CVC4::Command>* cmd]
   | CONTINUE_TOK
     { UNSUPPORTED("CONTINUE command"); }
   | RESTART_TOK formula[f] { UNSUPPORTED("RESTART command"); }
+  | RECURSIVE_FUNCTION_TOK (identifier[id,CHECK_NONE,SYM_VARIABLE] 
+    {
+      if(idCommaFlag){
+        idCommaFlag=false;
+      }
+      else{
+        PARSER_STATE->parseError("Identifiers need to be comma separated");
+      }
+    }
+    COLON type[t,CHECK_DECLARED] (COMMA {
+      idCommaFlag=true;
+      })? 
+    {
+      func = PARSER_STATE->mkVar(id, t, ExprManager::VAR_FLAG_NONE, true);
+      ids.push_back(id);
+      types.push_back(t);
+      funcs.push_back(func);
+    })+
+    EQUAL_TOK (formula[f]
+    {
+      if(formCommaFlag){
+        formCommaFlag=false;
+      }
+      else{
+        PARSER_STATE->parseError("Function definitions need to be comma separated");
+      }
+    }
+    (COMMA{
+      formCommaFlag=true;
+    })?
+    {
+      if( f.getKind()==kind::LAMBDA ){
+        for( unsigned i=0,size=f[0].getNumChildren(); i<size; i++)
+        {
+          bvs.push_back( f[0][i] );
+        }
+        formals.push_back(bvs);
+        bvs.clear();
+        f = f[1];
+        formulas.push_back(f);
+      }
+      else {
+        formals.push_back(bvs);
+        formulas.push_back(f);
+      }
+    })+
+    {
+      if(idCommaFlag){
+        PARSER_STATE->parseError("Cannot end function list with comma");
+      }
+      if(formCommaFlag){
+        PARSER_STATE->parseError("Cannot end function definition list with comma");
+      }
+      if(funcs.size()!=formulas.size()){
+        PARSER_STATE->parseError("Number of functions doesn't match number of function definitions");
+      }
+      for(unsigned int i = 0; i < funcs.size(); i++){
```

Comment:
Cache `funcs.size()`, e.g. `for(unsigned int i = 0, size = funcs.size(); i < size; i++)`